contrib/platform/src/com/sun/jna/platform/win32/PsapiUtil.java

```diff
@@ -0,0 +1,118 @@
+/* Copyright (c) 2020 Torbjörn Svensson, All Rights Reserved
+ *
+ * The contents of this file is dual-licensed under 2
+ * alternative Open Source/Free licenses: LGPL 2.1 or later and
+ * Apache License 2.0. (starting with JNA version 4.0.0).
+ *
+ * You can freely decide which license you want to apply to
+ * the project.
+ *
+ * You may obtain a copy of the LGPL License at:
+ *
+ * http://www.gnu.org/licenses/licenses.html
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "LGPL2.1".
+ *
+ * You may obtain a copy of the Apache License at:
+ *
+ * http://www.apache.org/licenses/
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+package com.sun.jna.platform.win32;
+
+import java.util.Arrays;
+
+import com.sun.jna.Memory;
+import com.sun.jna.Native;
+import com.sun.jna.platform.win32.WinDef.DWORD;
+import com.sun.jna.platform.win32.WinNT.HANDLE;
+import com.sun.jna.ptr.IntByReference;
+
+/**
+ * Psapi utility API.
+ *
+ * @author Torbj&ouml;rn Svensson, azoff[at]svenskalinuxforeninen.se
+ */
+public abstract class PsapiUtil {
+
+    /**
+     * Retrieves the process identifier for each process object in the system.
+     *
+     * @return Array of pids
+     */
+    public static int[] enumProcesses() {
+        int size = 0;
+        int[] lpidProcess = null;
+        IntByReference lpcbNeeded = new IntByReference();
+        do {
+            size = size + 1024;
+            lpidProcess = new int[size];
+            if (!Psapi.INSTANCE.EnumProcesses(lpidProcess, size * DWORD.SIZE, lpcbNeeded)) {
+                throw new Win32Exception(Native.getLastError());
+            }
+        } while (size == lpcbNeeded.getValue() / DWORD.SIZE);
+
+        return Arrays.copyOf(lpidProcess, lpcbNeeded.getValue() / DWORD.SIZE);
+    }
+
+    /**
+     * Retrieves the name of the process identified by <i>pid</i>
+     *
+     * @param pid The process identifier
+     * @return The name of the process
+     */
+    public static String getProcessName(int pid) {
+        HANDLE hProcess = null;
+        Win32Exception we = null;
+
+        try {
+            hProcess = Kernel32.INSTANCE.OpenProcess(WinNT.PROCESS_QUERY_INFORMATION | WinNT.PROCESS_VM_READ, false, pid);
+            if (hProcess == null) {
+                throw new Win32Exception(Native.getLastError());
+            }
+
+            return getProcessName(hProcess);
+        } catch (Win32Exception e) {
+            we = e;
+            throw we; // re-throw to invoke finally block
```

Comment:
I'm confused at the word "invoke" in this comment. The `finally` block will always be invoked.  I assume you mean passing the exception to the finally block somehow. If all this multiple-exception handling code is really needed I think it needs more explaining why.

However, I don't think it's possible to get multiple exceptions here. The only exception you'll get in the try block is the one you throw on a null handle. But you use the util version of closeHandle which doesn't throw an exception on null.

Reply:
I stole that part of the code from PsapiTest.testEnumProcessModules(), maybe it's wrong. I just guessed this was the construction to use in these cases. I am still puzzled about what it actually does.

Reply:
The thrown exception `e` would not be visible in the `finally` block, it would just be returned to the calling object.  Catching, assigning to an externally scoped variable and re-throwing it makes `we` accessible in the `finally` block.  IF you got a new exception it would "suppress" the old one, so adding it makes sure that the code will throw both any new exception you got plus `we`.

Except that'll never happen. :)

Reply:
This explains it better than I just did:  https://www.geeksforgeeks.org/throwable-addsuppressed-method-in-java-with-examples/

Reply:
I did not clean up this part in my last push. What should I do? Just remove the 'throw we' statement with the comment?

Reply:
I think you can remove the catch clause entirely, and simplify the finally clause to either just close the handle (and possibly throw any exception) or close the handle and catch any exception closing it.

Reply:
I tried some different scenarios, namely the following:

<pre>
 public class Main {
    public static void main(String[] args) {
        for (int i = 0; i <= 4; i++) {
            try {
                test(i);
            } catch (Exception e) {
                System.err.format("Test %d: ", i);
                e.printStackTrace();
            }
        }
    }

    private static void test(int tc) throws Exception {
        switch (tc) {
            case 0: {
                try {
                    throw new java.io.IOException("a");
                } finally {
                    throw new Exception("b");
                }
            }

            case 1:
            case 2: {
                Exception e = null;
                try {
                    throw new java.io.IOException("a");
                } catch (java.io.IOException ioe) {
                    e = ioe;
                    if (tc % 2 == 0) {
                        throw ioe;
                    }
                } finally {
                    throw new Exception("b");
                }
            }

            case 3:
            case 4: {
                Exception e = null;
                try {
                    throw new java.io.IOException("a");
                } catch (java.io.IOException ioe) {
                    e = ioe;
                    if (tc % 2 == 0) {
                        throw ioe;
                    }
                } finally {
                    try {
                        throw new java.io.IOException("b");
                    } catch (Exception e1) {
                        if (e != null) {
                            e.addSuppressed(e1);
                        } else {
                            e = e1;
                        }
                    }
                    if (e != null) {
                        throw e;
                    }
                }
                break;
            }
        }
    }
}
</pre>

<pre>
Test 0: java.lang.Exception: b
	at Main.test(Main.java:19)
	at Main.main(Main.java:5)
Test 1: java.lang.Exception: b
	at Main.test(Main.java:34)
	at Main.main(Main.java:5)
Test 2: java.lang.Exception: b
	at Main.test(Main.java:34)
	at Main.main(Main.java:5)
Test 3: java.io.IOException: a
	at Main.test(Main.java:42)
	at Main.main(Main.java:5)
	Suppressed: java.io.IOException: b
		at Main.test(Main.java:50)
		... 1 more
Test 4: java.io.IOException: a
	at Main.test(Main.java:42)
	at Main.main(Main.java:5)
	Suppressed: java.io.IOException: b
		at Main.test(Main.java:50)
		... 1 more
</pre>

What this shows is that test(0), test(1) and test(2) are equal regarding what exception is thrown.
test(3) and test(4) are also equal regarding what exception is thrown, but the exception in this case is the "inner" exception, in this case the exception thrown by Kernel32Util.QueryFullProcessImageName() rather than the Kernel32Util.closeHandle() call.

So to conclude, the `throw we;` statement could be removed, but not the rest if the error from Kernel32Util.QueryFullProcessImageName() is to be seen by the caller.

Reply:
The results for 0, 1, and 2 are exactly as expected: any exception thrown in the `finally` block will hide any other returned exception from `try` (if not caught) or rethrown in `catch`.  Similarly 3 and 4 are identical because in the catch block, `e` is always non-null so you always use the `e != null` branches, adding the suppressed second exception and then throwing the final exception in the `finally` block. (The `throw ioe` in the catch block never happens because you are always throwing `e` in the finally block.)

I do see your point that if `OpenProcess` succeeds (non-null `pHandle`) but `QueryFullProcessImageName` fails (generating an exception), that catch/save/rethrow logic would be necessary in the case we also got an exception closing the (non-null) handle. I missed that case in my earlier review.

So the code is fine as-is.

Reply:
> So the code is fine as-is.

For the application I'm working on, it does not matter, but I see that QueryFullProcessImageName does not completely support UNC. If I have a really long path to the binary, I still get the path returned as "C:\the\really\long\path\to\hello.exe" rather than the somewhat expected "\\?\C:\the\really\long\path\to\hello.exe". Obviously I've shortened the path for readability ;-)
The above is with the dwFlag=0. If I instead use dwFlag=1, I get "\Device\HarddiskVolume2the\really\long\path\to\hello.exe".
If I try the GetModuleFileNameEx version I originally pushed, I do get the UNC version.

In any case; should I include the fix for the path length issue of QueryFullProcessImageName(HANDLE,int) in this PR or simply extract the QueryFullProcessImageName(int,int) method to a new PR and leave this PR to only add the EnumProcesses?

Reply:
Up to you on whether to combine or update.  Regarding the UNC, have you tried `GetProcessImageFileName` to see if it returns something different than `QueryFullProcessImageName`?

Reply:
GetProcessImageFileName appears to give the same value as QueryFullProcessImageName with dwFlags=1.

As it would be easier to handle it in this PR, I will push again with a fixed QueryFullProcessImageName for long paths.